modules/ding_wayf_dk/modules/ding_wayf_dk_create_user/ding_wayf_dk_create_user.module

```diff
@@ -0,0 +1,397 @@
+    '#required' => TRUE,
+    '#attributes' => array(
+      'autocomplete' => 'off',
+    ),
+  );
+
+  $form['mail'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Email address'),
+    '#description' => t('The Email address that we can use to get in contact with you.'),
+    '#required' => TRUE,
+    '#attributes' => array(
+      'autocomplete' => 'off',
+    ),
+  );
+
+  // Get branch options form the active provider.
+  global $user;
+  $provider = _ding_provider_get_provider('reservation');
+  $options = ding_provider_invoke('reservation', 'options', $user);
+  $form['branch'] = $options[$provider['module'] . '_preferred_branch'];
+
+  $form['pin'] = array(
+    '#type' => 'password',
+    '#title' => t('Pin code'),
+    '#description' => t('Pin code with an max length of %length', array('%length' => ding_user_get_pincode_length())),
+    '#size' => 10,
+    '#maxlength' => ding_user_get_pincode_length(),
+    '#required' => TRUE,
+    '#attributes' => array(
+      'autocomplete' => 'off',
+    ),
+  );
+
+  $form['submit'] = array(
+    '#type' => 'submit',
+    '#value' => t('Create me'),
+  );
+
+  $link = '/wayf/logout';
+  if (user_is_logged_in()) {
+    $link = '/user/logout';
+  }
+  $form['cancel'] = array(
+    '#markup' => '<a class="ding-wayf--cancel" href="' . $link . '">' . t('Cancel') . '</a>',
+  );
+
+  $form['#submit'] = array('ding_wayf_dk_create_user_form_sbumit');
+
+  return $form;
+}
+
+/**
+ * Implements hook_wayf_dk_login_pre_logout().
+ *
+ * Ensures that session is cleaned up.
+ */
+function ding_wayf_dk_create_user_wayf_dk_login_pre_logout() {
+  if (isset($_SESSION['ding_wayf_dk_login_destination'])) {
+    unset($_SESSION['ding_wayf_dk_login_destination']);
+  }
+
+  if (isset($_SESSION['ding_wayf_dk_create_user'])) {
+    unset($_SESSION['ding_wayf_dk_create_user']);
+  }
+}
+
+/**
+ * User creation form submit handler.
+ */
+function ding_wayf_dk_create_user_form_sbumit($form, &$form_state) {
+  $values = $form_state['values'];
+  $cpr = $_SESSION['ding_wayf_dk_create_user'];
+
+  // Remove session variables used in redirects in the user creation process.
+  unset($_SESSION['ding_wayf_dk_create_user']);
+  unset($_SESSION['ding_wayf_dk_login_destination']);
+
+  // Try to create the user at the provider.
+  try {
+    // Create user.
+    ding_provider_invoke('user', 'create', $cpr, $values['pin'], $values['name'], $values['mail'], $values['branch']);
+
+    // Login the user and ensure he/her is redirect to notice page.
+    $_REQUEST['destination'] = 'wayf/create-user/notice';
+    ding_wayf_dk_login_wayf_dk_login_create_user(array(
+      'schacPersonalUniqueID' => array('urn:mace:terena.org:schac:personalUniqueID:dk:CPR:' . $cpr),
+    ));
+
+    // Remove request fields for the login form.
+    drupal_get_messages(NULL, TRUE);
+  }
+  catch (Exception $exception) {
+    drupal_set_message(t('You already have an user account at the library. So we have just logged you in.'));
```

Comment:
Does *any* exception thrown mean that the user already exist? That seems very suspicous.